partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelExclude.kt

```diff
@@ -0,0 +1,178 @@
+                it.step == structFieldCaseInsensitiveKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            // apply tuple wildcard exclusions
+            val tupleWildcardKey = ExcludeStep.StructWildcard
+            branches.find {
+                it.step == tupleWildcardKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            Pair(name, expr)
+        }
+        return structValue(finalStruct)
+    }
+
+    /**
+     * Returns a [PartiQLValue] created from an iterable of [coll]. Requires [type] to be a collection type
+     * (i.e. [PartiQLValueType.LIST], [PartiQLValueType.BAG], or [PartiQLValueType.SEXP]).
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    private fun newCollValue(type: PartiQLValueType, coll: Iterable<PartiQLValue>): PartiQLValue {
+        return when (type) {
+            PartiQLValueType.LIST -> listValue(coll)
+            PartiQLValueType.BAG -> bagValue(coll)
+            PartiQLValueType.SEXP -> sexpValue(coll)
+            else -> error("Collection type required")
+        }
+    }
+
+    @OptIn(PartiQLValueExperimental::class)
+    private fun excludeOnCollValue(
+        coll: CollectionValue<*>,
+        type: PartiQLValueType,
+        exclusions: ExcludeNode
+    ): PartiQLValue {
+        val leavesSteps = exclusions.leaves.map { leaf -> leaf.step }
+        val branches = exclusions.branches
+        if (leavesSteps.any { it is ExcludeStep.CollWildcard }) {
+            // collection wildcard at current level. return empty collection
+            return newCollValue(type, emptyList())
+        } else {
+            val indexesToRemove = leavesSteps.filterIsInstance<ExcludeStep.CollIndex>()
+                .map { it.index }
+                .toSet()
+            val collWithRemoved = when (coll) {
+                is BagValue -> coll
+                is ListValue, is SexpValue -> coll.filterIndexed { index, _ ->
+                    !indexesToRemove.contains(index)
+                }
+            }
+            val finalColl = collWithRemoved.mapIndexed { index, element ->
+                var expr = element
+                if (coll is ListValue || coll is SexpValue) {
+                    // apply collection index exclusions for lists and sexps
+                    val elementKey = ExcludeStep.CollIndex(index)
+                    branches.find {
+                        it.step == elementKey
+                    }?.let {
+                        expr = excludeOnPartiQLValue(element, it)
+                    }
+                }
+                // apply collection wildcard exclusions for lists, bags, and sexps
+                val collectionWildcardKey = ExcludeStep.CollWildcard
+                branches.find {
+                    it.step == collectionWildcardKey
+                }?.let {
+                    expr = excludeOnPartiQLValue(expr, it)
+                }
```

Comment:
Do you have an example of what this is for?

Reply:
I'll edit the comment to clarify. This basically applies all the collection wildcard exclude steps that have deeper exclude steps (e.g. `t.a.b.c[*].field_x` in
```
SELECT *
EXCLUDE
    t.a.b.c[*].field_x
FROM [{
    'a': {
        'b': {
            'c': [
                {                    -- c[0]; field_x to be removed
                    'field_x': 0, 
                    'field_y': 0
                },
                {                    -- c[1]; field_x to be removed
                    'field_x': 1,
                    'field_y': 1
                },
                {                    -- c[2]; field_x to be removed
                    'field_x': 2,
                    'field_y': 2
                }
            ]
        }
    }
}] AS t
```

When we get to the `c` list, the collection wildcard step will apply the remainder of the step (i.e. `.field_x`) on each of the elements of `c`.

Reply:
Ported above example to another test:

https://github.com/partiql/partiql-lang-kotlin/blob/50de0322624a06ecda07ea3c6bdfc1f3cefc5252/partiql-eval/src/test/kotlin/org/partiql/eval/internal/PartiQLEngineDefaultTest.kt#L249-L296